Improve WWA performance

srcarter3 opened the pull request:
AbstractWWAResource
- cache current frame
- cache current list of WWA entries
- add color, string, and DrawableString objects to the internal WarningEntry class
- use cached information when drawing, to draw faster (paintInternal method)

Commit:
Improve WWA performance

AbstractWWAResource
- cache current frame
- cache current list of WWA entries
- add color, string, and DrawableString objects to the internal WarningEntry class
- use cached information when drawing, to draw faster (paintInternal method)

tiffanycmeyer13 commented on cave/com.raytheon.viz.warnings/src/com/raytheon/viz/warnings/rsc/AbstractWWAResource.java:
```diff
@@ -420,47 +449,59 @@ protected void paintInternal(IGraphicsTarget target,
-        }
-        if (thisFrameTime == null) {
-            return;
-        }
```
Does this section need to be in here? It looks like just formatting changes

srcarter3 replied:
GitHub does a really poor job of the diff. A few sections are put into new conditionals (if statements) so that's why it looks like just a white space change. If you look closer you should be able to see the change.
Let me know if you want me to walk you through it?

Commit:
One more WWA change to go with the previous performance change
- text location needs to be recalulated with different zoom levels

tiffanycmeyer13 commented:
So the first time I load a WWA I get an error, but then I go load the same product it loads just fine. It happened once on my Windows VM (via eclipse) and once on a fresh install (direct Windows install):

```
ERROR 2023-03-29 11:39:18,648 0992 [main] CaveLogger: Paint error: null:: The resource [Flood Watches, Warnings, Advisories Wed 16:39Z 29-Mar-23] has been disabled.
com.raytheon.uf.viz.core.exception.VizException: Paint error: null:: The resource [Flood Watches, Warnings, Advisories Wed 16:39Z 29-Mar-23] has been disabled.
        at com.raytheon.uf.viz.core.drawables.AbstractRenderableDisplay.paintResource(AbstractRenderableDisplay.java:544)
        at com.raytheon.uf.viz.core.maps.display.MapRenderableDisplay.paint(MapRenderableDisplay.java:188)
        at com.raytheon.uf.viz.d2d.core.map.D2DMapRenderableDisplay.paint(D2DMapRenderableDisplay.java:172)
        at com.raytheon.viz.ui.panes.VizDisplayPane.glDrawInternal(VizDisplayPane.java:515)
        at com.raytheon.viz.ui.panes.VizDisplayPane.draw(VizDisplayPane.java:469)
        at com.raytheon.viz.ui.panes.DrawCoordinatedPane.draw(DrawCoordinatedPane.java:172)
        at com.raytheon.viz.ui.panes.DrawCoordinatorJob$1.run(DrawCoordinatorJob.java:194)
        at org.eclipse.ui.internal.PendingSyncExec.run(PendingSyncExec.java:68)
        at org.eclipse.ui.internal.UILockListener.doPendingWork(UILockListener.java:171)
        at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
        at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:185)
        at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:3987)
        at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3615)
        at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1157)
        at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
        at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
        at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
        at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:644)
        at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
        at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:551)
        at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:153)
        at com.raytheon.uf.viz.personalities.cave.component.CAVEApplication.startComponent(CAVEApplication.java:170)
        at com.raytheon.uf.viz.application.VizApplication.start(VizApplication.java:102)
        at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
        at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
        at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
        at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:401)
        at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:657)
        at org.eclipse.equinox.launcher.Main.basicRun(Main.java:594)
        at org.eclipse.equinox.launcher.Main.run(Main.java:1465)
Caused by: java.util.ConcurrentModificationException: null
        at java.base/java.util.HashMap$HashIterator.nextNode(HashMap.java:1511)
        at java.base/java.util.HashMap$ValueIterator.next(HashMap.java:1539)
        at com.raytheon.viz.warnings.rsc.AbstractWWAResource.paintInternal(AbstractWWAResource.java:505)
        at com.raytheon.uf.viz.core.rsc.AbstractVizResource.paint(AbstractVizResource.java:542)
        at com.raytheon.uf.viz.core.drawables.AbstractRenderableDisplay.paintResource(AbstractRenderableDisplay.java:541)
        ... 34 common frames omitted
```

Commit:
Additional small changes for WWA layer
- check the values() array for null, before attempting to cycle through it
- calculate the text positions the first time the display strings are created
- set the current zoom after drawing all the entries

srcarter3 commented:
I was not able to replicate the error you were seeing, but based on that stack trace, I *think* I know what the cause was, so I implemented a small change, that should address the issue.  If you could re-test it, that would be really helpful.
